src/core/ucc_context.c

```diff
@@ -0,0 +1,206 @@
+        ucc_cl_type_t *required_cls;
+        int            n_required_cls;
+        status = ucc_parse_cls_string(cls, &required_cls, &n_required_cls);
+        if (UCC_OK != status) {
+            ucc_error("failed to parse cls string: %s", cls);
+            return status;
+        }
+        for (i = 0; i < n_required_cls; i++) {
+            cl_cfg = find_cl_context_config(config, required_cls[i]);
+            if (!cl_cfg) {
+                ucc_error("required CL %s is not part of the context",
+                          ucc_cl_names[required_cls[i]]);
+                return UCC_ERR_INVALID_PARAM;
+            }
+            status = ucc_config_parser_set_value(
+                cl_cfg, cl_cfg->iface->cl_context_config.table, name, value);
+            if (UCC_OK != status) {
+                ucc_error("failed to modify CL \"%s\" configuration, name %s, "
+                          "value %s",
+                          cl_cfg->iface->super.name, name, value);
+                return status;
+            }
+        }
+        free(required_cls);
+    } else {
+        for (i = 0; i < config->n_cl_cfg; i++) {
+            if (config->configs[i]) {
+                status = ucc_config_parser_set_value(
+                    config->configs[i],
+                    config->lib->libs[i]->iface->cl_context_config.table, name,
+                    value);
+                if (UCC_OK != status) {
+                    ucc_error("failed to modify CL \"%s\" configuration, name "
+                              "%s, value %s",
+                              config->lib->libs[i]->iface->super.name, name,
+                              value);
+                    return status;
+                }
+            }
+        }
+    }
+    return UCC_OK;
+}
+
+void ucc_context_config_release(ucc_context_config_t *config)
+{
+    int i;
+    for (i = 0; i < config->n_cl_cfg; i++) {
+        if (!config->configs[i]) {
+            continue;
+        }
+        ucc_config_parser_release_opts(
+            config->configs[i],
+            config->lib->libs[i]->iface->cl_context_config.table);
+        free(config->configs[i]);
+    }
+    free(config->configs);
+    free(config);
+}
+
+/* The function prints the configuration of UCC context.
+   The ucc_context is a combination of contexts of different
+   (potentially multiple) CLs.
+
+   If HEADER flag is required - print it once passing user "title"
+   variable. For other cl_contexts use CL name as title so that
+   the printed output was clear for a user */
+void ucc_context_config_print(const ucc_context_config_h config, FILE *stream,
+                              const char *title,
+                              ucc_config_print_flags_t print_flags)
+{
+    int i;
+    int print_header = print_flags & UCC_CONFIG_PRINT_HEADER;
+    int flags        = print_flags;
+    /* cl_context_configs will always be printed with HEADER using
+       CL name as title */
+    flags |= UCC_CONFIG_PRINT_HEADER;
```

Comment:
@vspetrov I'm a bit confused here. Are you trying to print header for each context that belongs to different CL ?

Reply:
@manjugv There will a single header line printed as defined by the user (title argument). And then the name of CL printed as header will separate the output - it is more convenient, like sections of parameters for each CL. See example:

```
#
# USER TITLE HEADER
#



#
# CL_BASIC
#

#
# For dbg test purpuse : don't commit
#
# syntax:    unsigned integer
#
UCC_CL_BASIC_TEST_PARAM=5

#
# CL_OPT
#

#
# allreduce knomial radix
#
# syntax:    unsigned integer
#
UCC_CL_OPT_ALLREDUCE_KN_RADIX=4
```

The CL headers in the example above are : "CL_BASIC" and "CL_OPT".
We can remove this and just leave the list of params, they still have the prefix "UCC_<CL_NAME>" so stil can navigate. What do you think guys?